zebra-chain/src/parameters/network_upgrade.rs

```diff
@@ -223,6 +228,54 @@ impl NetworkUpgrade {
             }
         }
     }
+
+    /// Returns true if the gap between `block_time` and `previous_block_time` is
+    /// greater than the Testnet minimum difficulty time gap. This time gap
+    /// depends on the `network` and `block_height`.
+    ///
+    /// Returns false on Mainnet, when `block_height` is less than the minimum
+    /// difficulty start height, and when the time gap is too small.
+    ///
+    /// `block_time` can be less than, equal to, or greater than
+    /// `previous_block_time`, because block times are provided by miners.
+    ///
+    /// Implements the Testnet minimum difficulty adjustment from ZIPs 205 and 208.
+    ///
+    /// Spec Note: Some parts of ZIPs 205 and 208 previously specified an incorrect
+    /// check for the time gap. This function implements the correct "greater than"
+    /// check.
+    pub fn is_testnet_min_difficulty_block(
+        network: Network,
+        block_height: block::Height,
+        block_time: DateTime<Utc>,
+        previous_block_time: DateTime<Utc>,
+    ) -> bool {
+        let block_time_gap = block_time - previous_block_time;
+        if let Some(min_difficulty_gap) =
+            NetworkUpgrade::minimum_difficulty_spacing_for_height(network, block_height)
+        {
+            block_time_gap > min_difficulty_gap
+        } else {
+            false
+        }
+    }
+
+    /// Returns the averaging window timespan for the network upgrade.
+    ///
+    /// `AveragingWindowTimespan` from the Zcash specification.
+    pub fn averaging_window_timespan(&self) -> Duration {
+        self.target_spacing() * POW_AVERAGING_WINDOW as _
```

Comment:
oh wow, I'm surprised you can use `as` to convert into a `Duration`...

Reply:
It's the integer multiplier that's getting converted to the type that the Duration Mul impl accepts.

```suggestion
    pub fn averaging_window_timespan(&self) -> Duration {
        self.target_spacing() * (POW_AVERAGING_WINDOW as _)
```

Reply:
aah, makes sense.